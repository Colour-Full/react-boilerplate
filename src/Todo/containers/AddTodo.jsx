// @flow

import React, {Component} from 'react'
import { connect } from 'react-redux'
import { addTodo } from '../api-todo/actions/addTodo'
import type { Element } from 'react'
import { bindActionCreators, type Dispatch } from 'redux'
import Button from '@atlaskit/button'
import { Grid, GridColumn } from '@atlaskit/page'
import FieldText from '@atlaskit/field-text'

type State = {
  value: string
}
type AddTodoProps = {
  addTodo: (string) => any
}

export class AddTodo extends Component<AddTodoProps, State> {
  constructor (props: AddTodoProps) {
    super(props)
    this.state = { value: '' }

    this.handleChange = this.handleChange.bind(this)
    this.handleSubmit = this.handleSubmit.bind(this)
  }
  /*:: handleChange: (SyntheticEvent<HTMLInputElement>) => void */
  handleChange (event: SyntheticEvent<HTMLInputElement>) {
    this.setState({ value: event.currentTarget.value })
  }
  /*:: handleSubmit: (SyntheticEvent<HTMLFormElement>) => void */
  handleSubmit (event: SyntheticEvent<HTMLFormElement>) {
    event.preventDefault()
    if (!this.state.value.trim()) return

    this.props.addTodo(this.state.value)
  }

  render (): Element<any> {
    return (
      <form
        onSubmit={(event: SyntheticEvent<HTMLFormElement>): void => this.handleSubmit(event)}
      >
        <Grid>
          <GridColumn medium={4}>
            <FieldText type='text' value={this.state.value} onChange={this.handleChange} shouldFitContainer isLabelHidden />
          </GridColumn>
          <GridColumn medium={4}>
            <Button appearance='primary' type='submit'>
          Add To Do
            </Button>
          </GridColumn>
        </Grid>
      </form>
    )
  }
}

export const mapDispatchToProps = (dispatch: Dispatch<*>): any => bindActionCreators({
  addTodo
}, dispatch)

export default connect(null, mapDispatchToProps)(AddTodo)
