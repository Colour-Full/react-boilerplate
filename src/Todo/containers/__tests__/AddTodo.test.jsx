import React from 'react'
import { AddTodo } from '../AddTodo'
import Adapter from 'enzyme-adapter-react-16'
import Enzyme, { shallow } from 'enzyme'
import { shallowToJson } from 'enzyme-to-json'
import FieldText from '@atlaskit/field-text'

Enzyme.configure({ adapter: new Adapter() })

describe('AddTodo component', () => {
  it('Should render without throwing an error', () => {
    const addTodo = jest.fn()
    const wrapper = shallow(<AddTodo addTodo={addTodo} />)
    expect(shallowToJson(wrapper)).toMatchSnapshot()
  })
  it('Should update the state value property when the field text is changed', () => {
    const addTodo = jest.fn()
    const wrapper = shallow(<AddTodo addTodo={addTodo} />)
    wrapper.find(FieldText).simulate('change', { currentTarget: { value: 'Something to do' } })
    expect(wrapper.state().value).toEqual('Something to do')
  })
  it('Should call addTodo with the text field value when the form is submitted', () => {
    const addTodo = jest.fn()
    const wrapper = shallow(<AddTodo addTodo={addTodo} />)
    wrapper.find(FieldText).simulate('change', { currentTarget: { value: 'Something to do' } })
    wrapper.find('form').simulate('submit', {preventDefault: () => {}})
    expect(addTodo).toHaveBeenCalledWith('Something to do')
  })
  it('Should not call addTodo if text field value is empty', () => {
    const addTodo = jest.fn()
    const wrapper = shallow(<AddTodo addTodo={addTodo} />)
    wrapper.find(FieldText).simulate('change', { currentTarget: { value: '' } })
    wrapper.find('form').simulate('submit', { preventDefault: () => { } })
    expect(addTodo).toHaveBeenCalledTimes(0)
  })
})
