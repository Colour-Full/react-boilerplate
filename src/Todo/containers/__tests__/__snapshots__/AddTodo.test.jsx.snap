// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddTodo component Should render without throwing an error 1`] = `
<form
  onSubmit={[Function]}
>
  <WithTheme(AkGrid)>
    <WithTheme(AkGridColumn)
      medium={4}
    >
      <FieldText
        isLabelHidden={true}
        onChange={[Function]}
        shouldFitContainer={true}
        type="text"
        value=""
      />
    </WithTheme(AkGridColumn)>
    <WithTheme(AkGridColumn)
      medium={4}
    >
      <ForwardRef(WithAnalyticsContext(WithAnalyticsEvents(WithDeprecationWarnings(Button))))
        appearance="primary"
        type="submit"
      >
        Add To Do
      </ForwardRef(WithAnalyticsContext(WithAnalyticsEvents(WithDeprecationWarnings(Button))))>
    </WithTheme(AkGridColumn)>
  </WithTheme(AkGrid)>
</form>
`;
